javalib/src/main/scala/java/util/concurrent/Semaphore.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.util.concurrent
+
+import java.util.{Collection, Collections}
+
+class Semaphore(var permits: Int, fairness: Boolean) extends Serializable {
+
+  def this(permits: Int) = this(permits, false)
+
+  // These methods can't be implemented because calling it more times than the number
+  // of permits makes no sense in Javascript.
+  // def acquire(): Unit = ???
+  // def acquire(permits: Int): Unit = ???
+  // def acquireUninterruptibly(): Unit = this.acquire()
+  // def acquireUninterruptibly(permits: Int): Unit = this.acquire(permits)
+
+  def availablePermits(): Int = this.permits
+
+  def drainPermits(): Int = {
+    val permits = this.permits
+    this.permits = 0
+    permits
+  }
+
+  def getQueuedThreads(): Collection[Thread] = Collections.emptySet()
+
+  def getQueueLength(): Int = 0
+
+  def hasQueuedThreads(): Boolean = false
+
+  def isFair(): Boolean = this.fairness
+
+  protected def reducePermits(reduction: Int): Unit = {
+    this.permits = this.permits - reduction
+  }
+
+  def release(): Unit = this.release(1)
+
+  def release(permits: Int): Unit = {
+    if (permits < 0 ) {
+      throw new IllegalArgumentException()
+    }
+    this.permits += permits
+  }
+
+  override def toString: String =
+    s"${super.toString}[Permits = ${this.permits}]"
+
+  def tryAcquire(): Boolean = this.tryAcquire(1)
+
+  def tryAcquire(permits: Int): Boolean = {
+    if (permits > 0) {
+      this.permits -= permits
+      true
+    } else false
+  }
+
+  // This method is technically incorrect; it returns too soon
```

Comment:
If it is incorrect, it's better if we do not implement it. (this basically requires blocking to be implemented correctly, so maybe just add it to the list of methods we cannot implement at the top).